ops/model.py

```diff
@@ -230,6 +247,84 @@ def _get_unique(self, relation_name, relation_id=None):
             raise TooManyRelatedAppsError(relation_name, num_related, 1)
 
 
+class BindingMapping(Mapping):
+    """A map of binding names to lists of Binding instances."""
+
+    def __init__(self, binding_names, backend):
+        """
+        binding_names -- Relation endpoint names and names of extra-bindings.
+        backend -- An instance of a model backend implementation.
+        """
+        self._backend = backend
+        self._data = {binding_name: None for binding_name in binding_names}
+
+    def __contains__(self, key):
+        return key in self._data
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __getitem__(self, binding_name):
+        binding_list = self._data[binding_name]
+        if binding_list is None:
+            binding_list = self._data[binding_name] = []
+            # A relation id can be None which means a binding is not associated with a relation.
+            binding_list.append(Binding(binding_name, None, self._backend))
+            for relation_id in self._backend.relation_ids(binding_name):
+                binding_list.append(Binding(binding_name, relation_id, self._backend))
+        return binding_list
+
+    def _get_unique(self, binding_name, relation_id=None):
+        if not isinstance(relation_id, (int, type(None))):
+            raise ModelError(f'relation name {relation_id} must be int or None not {type(relation_id).__name__}')
+        for binding in self[binding_name]:
+            if binding.relation_id == relation_id:
+                return binding
+        raise ModelError(f'a Binding object named "{binding_name}" with a relation id "{relation_id}" does not exist')
+
+
+class Binding:
+    """A representation of a binding to a space."""
+
+    def __init__(self, name, relation_id, backend):
+        self.name = name
+        self.relation_id = relation_id
+        self._network_info = None
+        self._backend = backend
+
+    def _cache_network_info(self):
+        try:
+            if self.relation_id is None:
+                self._network_info = self._backend.network_get(self.name)
+            else:
+                self._network_info = self._backend.network_get(self.name, self.relation_id)
+        except RelationNotFoundError:
+            # If a relation is dead, we can still get network info associated with an endpoint itself
+            # not associated with a particular relation id.
+            self._network_info = self._backend.network_get(self.name)
+
+    @property
+    def network_info(self):
+        if not self._network_info:
+            self._cache_network_info()
+        return self._network_info
+
+    @property
+    def bind_address(self):
+        return self.network_info['bind-addresses'][0]['addresses'][0]['value']
```

Comment:
That's a pretty deep value just to find the default address. It also feels slightly unusual that this is a Binding with network_info["bind-addresses"][0]["addresses"] but bind_address is in the Binding and not in the network_info in the first place. We should probably talk for a moment about how to shape this object.

Meanwhile, where do I find docs about the format of the network-get result?

Reply:
Maybe we should just have NetworkInfo objects exposed as opposed to `Binding`s: https://github.com/canonical/operator/pull/109/files#r377494738

> Meanwhile, where do I find docs about the format of the network-get result?
https://jaas.ai/docs/charm-network-primitives#heading--examples

Reply:
Also on the format of the output, network-get simply writes all the data present in `NetworkInfoResults` type if no arguments requesting a specific piece of information are provided:

https://github.com/juju/juju/blob/juju-2.7.1/apiserver/params/network.go#L834-L846
```// NetworkInfoResult Adds egress and ingress subnets and changes the serialized
// `Info` key name in the yaml/json API protocol.
type NetworkInfoResult struct {
	Error            *Error        `json:"error,omitempty" yaml:"error,omitempty"`
	Info             []NetworkInfo `json:"bind-addresses,omitempty" yaml:"bind-addresses,omitempty"`
	EgressSubnets    []string      `json:"egress-subnets,omitempty" yaml:"egress-subnets,omitempty"`
	IngressAddresses []string      `json:"ingress-addresses,omitempty" yaml:"ingress-addresses,omitempty"`
}
```

https://github.com/juju/juju/blob/juju-2.7.1/api/uniter/unit.go#L810-L825
https://github.com/juju/juju/blob/juju-2.7.1/worker/uniter/runner/jujuc/network-get.go#L131-L134

Otherwise it retrieves the first address of a given type ("bind" or "ingress"):
https://github.com/juju/juju/blob/juju-2.7.1/worker/uniter/runner/jujuc/network-get.go#L152-L165